src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1420,16 +1430,83 @@ private IEnumerable<TypeInferenceInfo> InferTypeInImplicitArrayCreation(Implicit
                 return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
             }
 
+            private IEnumerable<TypeInferenceInfo> InferTypeInPropertyPattern(
+                PropertyPatternSyntax propertyPattern,
+                SyntaxNode child)
+            {
+                // something like: { ... sub patterns ... } or like: Blah { ... sub patterns ... }
+                // in the latter case, we know the type, because it's explicitly stated.  In the former
+                // we have to walk higher to figure it out.
+                if (propertyPattern.Type != null)
+                {
+                    return GetTypes(propertyPattern.Type);
+                }
+
+                return InferTypes(propertyPattern);
+            }
+
+            private IEnumerable<TypeInferenceInfo> InferTypeInPropertySubpattern(
+                PropertySubpatternSyntax propertySubpattern,
+                SyntaxNode child)
+            {
+                return InferTypes(propertySubpattern);
+            }
+
+            private IEnumerable<TypeInferenceInfo> InferTypeInSubpatternElement(
+                SubpatternElementSyntax subpatternElement,
+                SyntaxNode child)
+            {
+                // we have  { X: ... }.  The type of ... is whatever the type of 'X' is in its
+                // parent type.  So look up the parent type first, then find the X member in it
+                // and use that type.
+                if (child == subpatternElement.Pattern &&
+                    subpatternElement.NameColon.Name is IdentifierNameSyntax identifier)
```

Comment:
Note: we'd like to just ask what type 'X' has. However, that doesn't work in the SemanticModel currently.  @gafter When you add support in the semantic model, can you let me knwo, and i can update this to just use that approach.  Thanks!

Reply:
Yes. Also, I'd like to add a SemanticModel API that takes a PatternSyntax and tells you the input type coming in to that pattern.

---
In reply to: [183848025](https://github.com/dotnet/roslyn/pull/26355#discussion_r183848025) [](ancestors = 183848025)